Add infrastructure for safely spawning atoms at roundstart

optimumtact opened the pull request:
This system is called just before initialize and does not use a world
loop, so it is safe to spawn and create new atoms.

Any atom can inherit this behaviour by overriding the
spawn_atom_to_world proc and spawning atoms in this proc

Make sure you register the atom in question to the list of spawner atoms
by doing SSobj.atom_spawners += src in New(), and then remove when you are done
spawning objects

**Away missions - maploader etc:**

the same proc is called before initialize for all atoms loaded by the maploader, which should ensure that the logic works as expected on loaded maps and data.

Corpse spawners are moved to this system to prevent round start null errors.

Log showing roundstart errors are disappeared when using this system:
https://file.house/uJx1.log

optimumtact committed:
Add subsystem for safely spawning atoms at roundstart

This system is called just before initialize and does not use a world
loop, so it is safe to spawn and create new atoms.

Any atom can inherit this behaviour by overriding the
spawn_atom_to_world proc and spawning atoms there at roundstart

Make sure you register the atom in question to the list of spawner atoms
by doing SSobj.atom_spawners += src, and then remove when you are done
spawning objects

optimumtact commented on code/controllers/subsystem/objects.dm:
```diff
@@ -45,4 +52,5 @@ var/datum/subsystem/objects/SSobj
 		var/atom/B = A
-		B.initialize()
\ No newline at end of file
+		B.spawn_atom_to_world()
```
I've realised it makes more sense to just call trigger_atom_spawners() as it will have the list of items needing to spawn atoms

(assuming they add themselves to the processing list appropriately of course)

Menshin commented on code/modules/awaymissions/corpse.dm:
```diff
@@ -26,12 +26,16 @@
 /obj/effect/mob_spawn/New()
 	..()
+	//Add to the atom spawners register for roundstart atom spawning
+	SSobj.atom_spawners += src
```
```
if(roundstart)
    SSobj.atom_spawners += src
```
since this is supposed to be fired only at server init ?

optimumtact committed:
Minor tweak and fixes to subsystem logic

-only process roundstart spawners when they're needed

-use the same method to call atom_spawners for template loading.
    Assuming user is adding to the appropriate process list, this will not
cause issues

optimumtact commented:
There is a server running if people want to stress test this without lifting a finger

byond://jumphost.oranges.net.nz:45000

Core0verload commented:
I took even more hackish approach and spwan(0)'ed the corpse spawn out of the world loop.

optimumtact commented:
Do people consider it worth extending this to also process like the ssobjs.burning so people can spawn things during the round?

optimumtact commented:
This is ready to go once travis passes
